### PR TITLE
*: Fix issue#996 and handle prepared retry

### LINKED ISSUE #996: Insert auto_increment column with prepared statement failed
tidb> PREPARE stmt_1 FROM "insert into t (name) values (?);";
 Query OK, 0 row affected (0.00 sec) 
tidb> SET @t1="a", @t2="b", @t3="c"; 
Query OK, 0 row affected (0.00 sec)
 tidb> execute stmt_1 using @t1; 
2016/03/21 12:24:37 main.go:236: [error] github.com/pingcap/tidb/driver.go:59: query statement does not produce a result set (no top level SELECT) github.com/pingcap/tidb/driver.go:375:  github.com/pingcap/tidb/interpreter/main.go:65:   
tidb> execute stmt_1 using @t2;
2016/03/21 12:25:10 main.go:236: [error] github.com/pingcap/tidb/executor/executor_write.go:537: INSERT INTO t: expected 1 value(s), have 2 
github.com/pingcap/tidb/executor/executor_write.go:570:  github.com/pingcap/tidb/executor/executor_write.go:422:  github.com/pingcap/tidb/executor/adapter.go:84:  github.com/pingcap/tidb/tidb.go:172:  github.com/pingcap/tidb/session.go:376:  github.com/pingcap/tidb/driver.go:372:  github.com/pingcap/tidb/interpreter/main.go:65:

### PR DESCRIPTION
**Pull request by zimulala**
fix #996

### CODE REVIEW

**Comment by coocood** [comment]
LGTM

File: sessionctx/variable/session.go [hunk 1]
```diff
@@ -17,6 +17,33 @@ import (
 	"github.com/pingcap/tidb/context"
 )
 
+// RetryInfo saves retry infomation.
+type RetryInfo struct {
+	Retrying         bool
+	CurrRetryOff     int
```
Comment (shenli) [inline sessionctx/variable/session.go]: Change this a private attribute. Use Reset() interface to reset this cursor.

File: sessionctx/variable/session.go [hunk 2]
```diff
@@ -17,6 +17,33 @@ import (
...
+
+// AddAutoIncrementID adds id to AutoIncrementIDs.
+func (r *RetryInfo) AddAutoIncrementID(id int64) {
+	r.autoIncrementIDs = append(r.autoIncrementIDs, id)
+}
+
+// GetCurrAutoIncrementID gets current AutoIncrementID.
+func (r *RetryInfo) GetCurrAutoIncrementID() int64 {
+	id := r.autoIncrementIDs[r.CurrRetryOff]
+	r.CurrRetryOff++
```
Comment (shenli) [inline sessionctx/variable/session.go]: If CurrRetryOff >= len(autoIncrementIDs), return an error.

**Comment by shenli** [comment]
LGTM

**Pull request merged**